src/backend/src/lib.rs

```diff
@@ -1,33 +1,143 @@
+    DefaultMemoryImpl, StableBTreeMap, StableCell,
+};
 use k256::PublicKey;
 use serde_bytes::ByteBuf;
 use shared::http::{HttpRequest, HttpResponse};
 use shared::metrics::get_metrics;
 use shared::std_canister_status;
+use std::borrow::Cow;
 use std::cell::RefCell;
 use std::str::FromStr;
 
 mod guards;
 
+type VMem = VirtualMemory<DefaultMemoryImpl>;
+type ConfigCell = StableCell<Option<Candid<Config>>, VMem>;
+type ChainId = u64;
+type UserTokenMap = StableBTreeMap<StoredPrincipal, Candid<Vec<Token>>, VMem>;
+
+const CONFIG_MEMORY_ID: MemoryId = MemoryId::new(0);
+const USER_TOKEN_MEMORY_ID: MemoryId = MemoryId::new(1);
+
+const MAX_SYMBOL_LENGTH: usize = 20;
+const MAX_TOKEN_LIST_LENGTH: usize = 100;
+
 thread_local! {
-    static STATE: RefCell<Option<State>> = RefCell::default();
+    static MEMORY_MANAGER: RefCell<MemoryManager<DefaultMemoryImpl>> = RefCell::new(
+        MemoryManager::init(DefaultMemoryImpl::default())
+    );
+
+    static STATE: RefCell<State> = RefCell::new(
+        MEMORY_MANAGER.with(|mm| State {
+            config: ConfigCell::init(mm.borrow().get(CONFIG_MEMORY_ID), None).expect("config cell initialization should succeed"),
+            user_token: UserTokenMap::init(mm.borrow().get(USER_TOKEN_MEMORY_ID)),
+        })
+    );
 }
 
 pub fn read_state<R>(f: impl FnOnce(&State) -> R) -> R {
-    STATE.with(|cell| f(cell.borrow().as_ref().expect("state not initialized")))
+    STATE.with(|cell| f(&*cell.borrow()))
+}
+
+pub fn mutate_state<R>(f: impl FnOnce(&mut State) -> R) -> R {
+    STATE.with(|cell| f(&mut *cell.borrow_mut()))
+}
+
+pub fn read_config<R>(f: impl FnOnce(&Config) -> R) -> R {
+    read_state(|state| {
+        f(state
+            .config
+            .get()
+            .as_ref()
+            .expect("config is not initialized"))
+    })
+}
+
+#[derive(Default)]
+struct Candid<T>(T)
+where
+    T: CandidType + for<'de> Deserialize<'de>;
+
+impl<T> Storable for Candid<T>
+where
+    T: CandidType + for<'de> Deserialize<'de>,
+{
+    const BOUND: Bound = Bound::Unbounded;
+
+    fn to_bytes(&self) -> Cow<'_, [u8]> {
+        Cow::Owned(candid::encode_one(&self.0).expect("encoding should always succeed"))
+    }
+
+    fn from_bytes(bytes: Cow<'_, [u8]>) -> Self {
+        Self(candid::decode_one(bytes.as_ref()).expect("decoding should succeed"))
+    }
+}
+
+impl<T> Deref for Candid<T>
+where
+    T: CandidType + for<'de> Deserialize<'de>,
+{
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        &self.0
+    }
 }
 
-#[derive(CandidType, Deserialize)]
 pub struct State {
+    config: ConfigCell,
+    user_token: UserTokenMap,
+}
+
+#[derive(CandidType, Deserialize)]
+pub struct Token {
+    pub contract_address: String,
+    pub chain_id: ChainId,
+    pub symbol: Option<String>,
+    pub decimals: Option<u8>,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+struct StoredPrincipal(Principal);
+
+impl Storable for StoredPrincipal {
```

Comment:
Out of curiosity, doesn't unbounded tree map handle automatically principal now?
i.e. this implementation `StoredPrincipal` is a must?

Reply:
Principals aren't storable, unbounded keys/values are orthogonal to the encoding.

Reply:
Weird. [Here](https://github.com/junobuild/juno/blob/1f4043de026e0e04d6731f3c91ce550d737a0c41/src/satellite/src/db/impls.rs#L19) I implement storable for struct that contains a principal and I do not get any errors, at least at build time. I should double check then at runtime.

Reply:
Thanks for the feedback.

Reply:
Ah it's because the principal is used within the value, not the key!
Sorry for the noise.